src/datastore/src/filestore.js

```diff
@@ -232,164 +232,239 @@ export default class FileStore extends NetworkStore {
-              return file;
-            }
-
-            // Upload the file
-            options.start = getStartIndex(statusCheckResponse.headers.get('range'), metadata.size);
-            return this.uploadToGCS(uploadUrl, headers, file, metadata, options);
-          })
-          .then((file) => {
-            data._data = file;
-            return data;
-          });
+    return this.saveFileMetadata(options, metadata)
+      .then((response) => {
+        kinveySaveData = response.data;
+        return this.makeStatusCheckRequest(response.data, metadata, options.timeout);
+      })
+      .then((response) => {
+        Log.debug('File upload status check response', response);
+
+        if (response.isSuccess() === false) {
+          return Promise.reject(response.error);
+        }
+
+        if (response.statusCode !== 308) {
+          // TODO: Here we should handle redirects according to location header, but this generally shouldn't happen
+          const err = this.getUnexpectedStatusCheckResponsCode(response);
+          return Promise.reject(err);
+        }
+
+        const uploadOptions = {
+          start: getStartIndex(response.headers.get('range'), metadata.size),
+          timeout: options.timeout,
+          maxBackoff: options.maxBackoff,
+          headers: kinveySaveData._requiredHeaders
+        };
+        return this.retriableUpload(kinveySaveData._uploadURL, file, metadata, uploadOptions);
+      })
+      .then(() => {
+        return this.formUploadResponse(kinveySaveData, file);
       });
   }
 
   /**
    * @private
    */
-  uploadToGCS(uploadUrl, headers, file, metadata, options = {}) {
-    // Set default options
+  resolvePromiseAfter(timeout, resolveWith) {
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        resolve(resolveWith);
+      }, timeout || 0);
+    });
+  }
+
+  /**
+   * @private
+   */
+  handleUploadErrors(uploadResponse) {
+    if (uploadResponse.isClientError()) {
+      return Promise.reject(uploadResponse.error);
+    }
+    if (!uploadResponse.isSuccess() && !uploadResponse.isServerError() && uploadResponse.statusCode !== 308) {
+      // TODO: Here we should handle redirects according to location header
+      return Promise.reject(this.getUnexpectedStatusCheckResponsCode(uploadResponse));
+    }
+    return Promise.resolve(uploadResponse);
+  }
+
+  /**
+   * @private
+   * This function may mutate the options object param - may change its start property
+   */
+  handleReuploadCases(uploadResponse, fileSize, options) {
+    let backoff = 0;
+    const reuploadData = {
+      shouldResume: false,
+      shouldRetry: false
+    };
+
+    if (uploadResponse.isServerError()) { // should retry
+      Log.debug('File upload server error. Probably network congestion.', uploadResponse.statusCode, uploadResponse.data);
+      backoff = (2 ** options.count) + (1000 + randomInt(-100, 101)); // Calculate the exponential backoff
```

Comment:
I rewroked this a bit, since the way it used to be was more random than it was exponential - you could get 1x1000, then 2x10, then 4x1000, 8x2. This version isn't all that mathematically proven, but I think it's a good balance between exponential with a jitter and simplicity :)